sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -279,28 +290,31 @@ public void processElement(ProcessContext processContext) throws Exception {
       while (!finished) {
         ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
         List<String> keys = scanResult.getResult();
-
-        Pipeline pipeline = jedis.pipelined();
-        if (keys != null) {
-          for (String key : keys) {
-            pipeline.get(key);
-          }
-          List<Object> values = pipeline.syncAndReturnAll();
-          for (int i = 0; i < values.size(); i++) {
-            processContext.output(KV.of(keys.get(i), (String) values.get(i)));
-          }
+        for (String k : keys) {
+          processContext.output(k);
         }
-
         cursor = scanResult.getStringCursor();
         if ("0".equals(cursor)) {
           finished = true;
         }
       }
     }
+  }
+  /** A {@link DoFn} requesting Redis server to get key/value pairs. */
+  private static class ReadFn extends BaseReadFn<KV<String, String>> {
 
-    @Teardown
-    public void teardown() {
-      jedis.close();
+    ReadFn(RedisConnectionConfiguration connectionConfiguration) {
+      super(connectionConfiguration);
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext processContext) throws Exception {
+      String key = processContext.element();
+
+      String value = jedis.get(key);
```

Comment:
As mentioned in the previous PR I am a bit concerned about losing the multiple data request capability, any chance you can work on this with the approach you mentioned based on `MGET` for ReadFn.
The simplest approach probably is to do like other IOs and have a default size that can be parametrized via a `withBatchSize` method. WDYT ?

Reply:
Sure, I think it should be straightforward, will update the pr over the week.

Reply:
@iemejia right now there are two operators exposed for read operations 
1. ReadKeysWithPattern - which is 2 stepped - a ) for each pattern prefix in the PCollection fetch all matching keys b) fetch values for each key 
2. ReadFn - which gets values for each key in the PCollection.

I can make changes to ReadKeysWithPattern to use a parameterized batch to fetch the data in the b step, but for ReadFn I am not sure of how to use the batch parameter.

Reply:
Hi, sorry I have missed your message. The idea is that we should add the DoFn startBundle and finishBundle methods and create a method in the Read to define the size of the maximum amount of elements that we will request, then you will build the collection of the keys that are going to be requested in the processElement, but you  won't do the request in the processElement but in the finishBundle method by doing a MGET request with the defined number of elements of the batch, we should choose a default min size e.g. 1000. It is similar to what other IOs do in the Write (see withBatchSize in ElasticsearchIO or SolrIO, for ref.
https://github.com/apache/beam/blob/c14c975224af417dcdc74fed8b0d893be742e9d7/sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java#L805-L829